### ISSUE TITLE
prefer-destructuring false Positive when casting result

### ISSUE THREAD

**Issue by sk-** [eslint/eslint#13678]
**Tell us about your environment**

```
Environment Info:

Node version: v14.6.0
npm version: v6.14.7
Local ESLint version: v7.8.1 (Currently used)
Global ESLint version: Not found
```

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**
default
**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
"prefer-destructuring": ["error", {"VariableDeclarator": {"object": true}}],
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
const transactionData =
      /** @type {Transaction} */
      (data.transactionData);
```

```bash

```

**What did you expect to happen?**
Not to complain about destructuring

**What actually happened? Please include the actual, raw output from ESLint.**
```
error  Use object destructuring  prefer-destructuring
```

**Are you willing to submit a pull request to fix this bug?**
Depending on whether I'm pointed in the right direction.

**Comment by mdjermanovic** [comment]
Hi @sk-, thanks for the bug report!

I'm not sure if this is a bug, but I think it's reasonable to support this use case, either by default or behind an option.

Maybe we could check if there are any comments between the `=` and the right side, and not report an error in such a case.

There's also a bug that should be certainly fixed: autofix removes that comment, which should never happen. At least the rule shouldn't auto-fix this code.

**Comment by mdjermanovic** [comment]
Would the following be equivalent to the original code?

```js
const { transactionData } =
      /** @type {{ transactionData: Transaction }} */
      (data);
```

**Comment by sk-** [comment]
@mdjermanovic yes, that works, but in that case we lose all the benefits of destructuring, as `transactionData` is now also repeated. Compare:

```js
const transactionData =
      /** @type {Transaction} */
      (data.transactionData);
```

vs 

```js
const { transactionData } =
      /** @type {{ transactionData: Transaction }} */
      (data);
```

The second option is even more verbose, as there are 2 extra `{}` pairs, plus the extra whitespace.

**Comment by mdjermanovic** [comment]
Thanks for clarifying! Then, I think this isn't a bug, but I'll support a configuration option to not report an error if there are any comments inside the declaration/assignment. Let's see if this enhancement will reach a [consensus](https://eslint.org/docs/maintainer-guide/issues#accepting-issues) from the team.

I'm working on the autofix bug separately.

**Comment by anikethsaha** [comment]
Is it a stylistic rule? it is in the category of es6 but It does works for stylistic preference I suppose. 
If it is a stylistic rule then supporting this check will fall under [new policies](https://eslint.org/blog/2020/05/changes-to-rules-policies). 

If not, IMO, it should be in [jsdoc-plugins](https://github.com/gajus/eslint-plugin-jsdoc) right ? cause js doesn't understand that typecasting using comments. 


WDYT ?
